dashboard/app/controllers/projects_controller.rb

```diff
@@ -133,10 +133,59 @@ def public
     end
   end
 
+  # @param [Array {Hash}] Each hash is data for a row in the featured projects tables.
+  # The rows are sorted into two arrays: featured or unfeatured, based on
+  # on whether the project is currently featured or not.
+  def sort_projects(featured_project_table_rows)
+    @featured = []
+    @unfeatured = []
+    featured_project_table_rows.each do |row|
+      row[:featured] ? @featured << row : @unfeatured << row
+    end
+  end
+
+  # @param [Hash] key is a FeaturedProject, value is the associated project
+  # Iterate through the Featured Project - Project pairs are creates a new hash # for each containing combined information specific to what the featured
+  # projects tables are expecting.
+  def combine_projects_and_featured_projects_data(matched_projects_and_featured_projects)
+    @featured_project_table_rows = []
+    matched_projects_and_featured_projects.each do |featured_project, project|
+      project_details = JSON.parse(project[:id][:value])
+      featured_project_row = {
+        projectName: project_details['name'],
+        type: project_details['projectType'],
+        publishedAt: project_details['publishedAt'],
+        thumbnailUrl: project_details['thumbnailUrl'],
+        featuredAt: featured_project.featured_at,
+        unfeaturedAt: featured_project.unfeatured_at,
+        featured: featured_project.featured?
+      }
+      @featured_project_table_rows << featured_project_row
+    end
+    sort_projects(@featured_project_table_rows)
+  end
+
+  # @param featured_projects [Array of FeaturedProjects]
+  # Iterates through featured_projects and finds the project associated with each FeaturedProject
+  def fetch_projects(featured_projects)
+    @matched_projects_and_featured_projects = {}
+    featured_projects.each do |featured_project|
+      @matched_projects_and_featured_projects[featured_project] = PEGASUS_DB[:storage_apps].where(id: featured_project.storage_app_id)
```

Comment:
This does feel like a it generates a lot of read queries, especially for an unlimited number of archived projects.

I think we might want to limit the number of archived projects we show, or at least explicitly ask for them? Maybe we paginate, or add a search?

Reply:
I'm new to the ruby world, so I don't have great guidance on how to change this, but @davidsbailey I'm curious about your thoughts.

Reply:
this should be done with a join, so that we fetch all the data in one SQL query. it would look something like this: https://github.com/code-dot-org/code-dot-org/blob/418fb5b8d0cbfe4e63cdc2fa6ed2bea74b14b543/dashboard/lib/projects_list.rb#L98-L128